tablo/api.py

```diff
@@ -403,11 +405,26 @@ def describe(self, request, **kwargs):
         if obj.extension == 'csv':
             csv_file_name = obj.file.name
         else:
-            raise ImmediateHttpResponse(HttpBadRequest('Unsupported file format.'))
+            error_json = derive_error_response_data(InvalidFileError(
+                underlying='Unsupported file format', extension=obj.extension
```

Comment:
Very good points. I made a few changes with a mind to improve what is happening here in order to:

1. Improve this pattern so that it is more intuitively extensible
2. Prevent the instantiation of exceptions that are never raised
3. Continue to encapsulate error-specific data in properties of custom exceptions
4. Continue to encapsulate JSON data output in `derive_error_response_data`
5. Not add TastyPie-specific request/response behavior or classes to `tablo.exceptions`

Overall, the purpose of `derive_error_response_data` is to return uniform error response data from an exception. It is needed because many times we have nothing but a raw Exception message coming from Postgres through `django.db.utils.DatabaseError`, but we need to provide something useful to the client (Data Basin in this case). I want to leverage it here so that business logic handling exception output is managed in one place, in `tablo.exceptions`, whether the logic itself fits better in our custom Exceptions, or in the handling of custom or 3rd party exceptions after they are raised.

This is what I did:

1. Updated `derive_error_response_data` to take a code as you suggested (though it defaults instead to `UNKNOWN_ERROR`)
2. Updated the two blocks so that they detect a bad file condition before raising a built-in exception, and instead raise our custom `InvalidFileError`
3. In the `except` blocks for each place this is done, call `derive_error_response_data` in a standard way